### PR TITLE
chore: drop support for Python 2

### LINKED ISSUE #2896: Python 2.7 support
## Issue

- [x] This is not a bug report, feature request, or plugin issue/request.
- [x] I have read the contribution guidelines.

### Description

Hi ! 
This is more of a question and the beginning of a discussion than an issue.

Python 2.7 is not maintained anymore (see: https://github.com/python/devguide/pull/344)

It has been mentioned a few times in the project's issues that Python 2.7 is sometimes problematic.
Last time was only a few days ago (see: https://github.com/streamlink/streamlink/pull/2871#issuecomment-611939149)

I'm really new in this project and I understand that there are historical reasons for a Python 2.7 version of StreamLink but I would like to know if there is a plan on Python 2.7 support.

Maybe some users are still relying on Python 2.7, should they expect an end of support soon?
How many users would be impacted? Is there an easy way to know?

### PR DESCRIPTION
**Pull request by bastimeyer**
- Bump Python version requirement to 3.5
- Remove unneeded compatibility dependencies
- Remove py2 warning from streamlink.session
- Update pypi package classifiers
- Update docs
- Remove universal bdist_wheel
- Remove py2 from CI test runners

----

This is **WIP** and needs to be properly tested in a test-release first.

Additional resources:
- https://packaging.python.org/guides/distributing-packages-using-setuptools/#universal-wheels
- https://packaging.python.org/guides/distributing-packages-using-setuptools/#classifiers
- https://pypi.org/classifiers/

Resolves #2896 
Closes #2920

### CODE REVIEW

**Comment by bastimeyer** [comment]
```
$ python --version
Python 2.7.18
$ pip install -e .
DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
Obtaining file:///home/basti/repos/streamlink
ERROR: Package 'streamlink' requires a different Python: 2.7.18 not in '>=3.5, <4'
```

**Comment by bastimeyer** [comment]
Next steps (in separate PRs) after this has been merged:
1. **Remove compat stuff**
   This involves all imports from the compat modules and also custom `if is_py2` checks, etc. This can't be automated.
2. **Remove py2 specifics and translate code into "modern" py3 code**
   Most of it can be automated, but not everything. This is all optional, but should be done eventually. With the removal of py35, even more new syntax features become available, but that should be a separate step.
   The automation tool used in one of the other PRs unfortunately hardcodes the features it translates into py3, so before running it, the unwanted stuff needs to be disabled first.

Oh, and I'd recommend having a 1.6.1 or 1.7.0 release before bumping to 2.0.0

**Comment by bastimeyer** [comment]
I have the compatibility removal branch in a somewhat ready state here:
https://github.com/bastimeyer/streamlink/compare/py2-removal...bastimeyer:py2-compat-removal

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ jobs:
       # when changing the build matrix and changing the number of test runners
       matrix:
         os: [ubuntu-latest, windows-latest]
-        python: [2.7, 3.5, 3.6, 3.7, 3.8, 3.9]
+        python: [3.5, 3.6, 3.7, 3.8, 3.9]
```
Comment (back-to) [inline .github/workflows/main.yml:24]: you can also drop **3.5** support
https://devguide.python.org/#status-of-python-branches

File: setup.cfg [hunk 2]
```diff
@@ -1,6 +1,3 @@
-[wheel]
```
Comment (back-to) [inline setup.cfg:1]: might be better

```
[bdist_wheel]
python-tag = py36.py37.py38.py39
```

Comment (bastimeyer) [reply setup.cfg:1]: That would rename the built wheel from
`streamlink-1.6.0+31.g4cd86ba-py3-none-any.whl`
to
`streamlink-1.6.0+31.g4cd86ba-py36.py37.py38.py39-none-any.whl`

On the universal wheel, the python tag was just `py2.py3` and didn't include specific py3 versions:
`streamlink-1.6.0-py2.py3-none-any.whl`

File: .github/workflows/main.yml (hunk 1 above)
Comment (bastimeyer) [reply .github/workflows/main.yml:24]: Yes, I know, I wanted to do this afterwards.

**Review by back-to — APPROVED** [review]

**Review by gravyboat — APPROVED** [review]
Thumbs up from me, looks good!

**Review by beardypig — APPROVED** [review]

**Comment by bastimeyer** [comment]
Squashed both commits and updated the commit message accordingly.
Going to merge now, as everyone has approved. :tada:

**Pull request merged**